std-bits/base/src/main/java/org/enso/base/enso_cloud/TransientHTTPResponseCache.java

```diff
@@ -0,0 +1,419 @@
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import org.apache.commons.io.IOUtils;
+import org.enso.base.net.URIWithSecrets;
+import org.graalvm.collections.Pair;
+
+/**
+ * TransientHTTPResponseCache is a cache for EnsoHttpResponse values that respects the cache control
+ * HTTP headers received in the original repsonse to a request.
+ *
+ * <p>It also puts limits on the size of files that can be requested, and on the total cache size,
+ * deleting entries to make space for new ones.
+ *
+ * <p>Without cachineg, EnsoHttpResponse contains an InputStream providing the response data. When
+ * there is a cache hit, this stream reads from the local file storing the cached data. When there
+ * is no cache hit, the InputStream is connected directly to the remote server.
+ */
+public class TransientHTTPResponseCache {
+  TransientHTTPResponseCache() {}
+
+  private static final Logger logger = Logger.getLogger(TransientHTTPResponseCache.class.getName());
+
+  // 1 year.
+  private static final int DEFAULT_TTL_SECONDS = 31536000;
+
+  private static long MAX_FILE_SIZE = 10L * 1024 * 1024;
+  private static long MAX_TOTAL_CACHE_SIZE = 10L * 1024 * 1024 * 1024;
+
+  /** This value is used for the current time when deciding if a cache entry is stale. */
+  private static ZonedDateTime nowOverrideTestOnly = null;
+
+  /**
+   * Used for testing file and cache size limits. These cannot be set to values larger than the real
+   * limits.
+   */
+  private static Optional<Long> maxFileSizeOverrideTestOnly = Optional.empty();
+
+  private static Optional<Long> maxTotalCacheSizeOverrideTestOnly = Optional.empty();
+
+  private static final Map<String, CacheEntry> cache = new HashMap<>();
+  private static final Map<String, ZonedDateTime> lastUsed = new HashMap<>();
+
+  static EnsoHttpResponse makeRequest(
+      URIWithSecrets unresolvedURI,
+      URI resolvedURI,
+      List<Pair<String, String>> resolvedHeaders,
+      RequestMaker requestMaker)
+      throws IOException, InterruptedException, ResponseTooLargeException {
+    var cacheKey = makeHashKey(resolvedURI, resolvedHeaders);
+    if (cache.containsKey(cacheKey)) {
+      return returnCachedResponse(cacheKey);
+    } else {
+      return makeRequestAndCache(
+          unresolvedURI, resolvedURI, resolvedHeaders, cacheKey, requestMaker);
+    }
+  }
+
+  private static EnsoHttpResponse returnCachedResponse(String cacheKey) throws IOException {
+    markCacheEntryUsed(cacheKey);
+    return buildEnsoHttpResponseFromCacheEntry(cache.get(cacheKey));
+  }
+
+  /**
+   * IOExceptions thrown by the HTTP request are propagated; IOExceptions thrown while storing the
+   * data in the cache are caught.
+   *
+   * <p>Only HTTP 200 responses are cached; all others are returned uncached.
+   */
+  private static EnsoHttpResponse makeRequestAndCache(
+      URIWithSecrets unresolvedURI,
+      URI resolvedURI,
+      List<Pair<String, String>> resolvedHeaders,
```

Comment:
IDE tells me that this argument is unused. It may contain secrets, so we need to be very careful about passing it around. Let's remove the unused arg.